contracts/service-provider-directory/src/contract/execute.rs

```diff
@@ -74,47 +75,98 @@ fn return_deposit(service_to_delete: &Service) -> BankMsg {
     }
 }
 
+fn verify_announce_signature(
+    deps: Deps<'_>,
+    sender: Addr,
+    deposit: Coin,
+    service: ServiceDetails,
+    signature: MessageSignature,
+) -> Result<()> {
+    // recover the public key
+    let public_key = decode_ed25519_identity_key(&service.identity_key)?;
+
+    // reconstruct the payload
+    let nonce = state::get_signing_nonce(deps.storage, sender.clone())?;
+
+    let msg = construct_service_provider_announce_sign_payload(nonce, sender, deposit, service);
+
+    if deps.api.verify_message(msg, signature, &public_key)? {
+        Ok(())
+    } else {
+        Err(SpContractError::InvalidEd25519Signature)
```

Comment:
Another thing we had to redefine, maybe at some point (not this PR!) we should move signing helpers and associated errors to separate crate?